File: examples/exo/cpp/src/discover_exos.cpp
#include "exo_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>

int main() {
    try {
        for (const auto& device : qnbot::discover_exos()) {
            if (device.error) {
                std::cout << "port=" << device.port
                          << " error=" << device.error->message << '\n';
                continue;
            }
            if (!device.device_info) continue;
            const auto& info = *device.device_info;
            std::cout << "port=" << device.port
                      << " product=" << static_cast<int>(info.product)
                      << " protocol="
                      << static_cast<int>(info.protocol_version.major) << '.'
                      << static_cast<int>(info.protocol_version.minor);
            if (info.serial_number) std::cout << " sn=" << *info.serial_number;
            if (info.hardware_name)
                std::cout << " name=" << *info.hardware_name;
            std::cout << '\n';
        }
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: examples/exo/python/src/discover_exos.py
from __future__ import annotations

from qnbot_sdk.exo import discover_exos


def main() -> None:
    for device in discover_exos():
        if device.error is not None:
            print(
                f"port={device.port} error={device.error.code.value} "
                f"message={device.error.message}"
            )
            continue
        info = device.device_info
        print(
            f"port={device.port} product={info.product.name} "
            f"protocol={info.protocol_version.major}.{info.protocol_version.minor} "
            f"sn={info.serial_number} name={info.hardware_name}"
        )


if __name__ == "__main__":
    main()
